### DISCUSSION TITLE
Usage With Unity Project for Android

### DISCUSSION THREAD

**Discussion by jakedowns** [Tyrrrz/YoutubeExplode#679] [Q&A] [answered]
Hello, I'm trying to use this wonderful package with a Unity 2021.3 project,
but i'm getting this error:

`ExecutionEngineException: Attempting to call method 'System.Text.Json.Serialization.Converters.SmallObjectWithParameterizedConstructorConverter`

https://answers.unity.com/questions/1934842/android-executionengineexception-systemtextjsonser.html

when i install the package from NuGet, i then copy the .dll files out of `/Packages/.../lib/netstandard2.0/*.dll` and move them to `/Assets/Plugins/YoutubeExplode/*`

anyone gotten this to work with Unity successfully in the past?

**Comment by Tyrrrz** [comment]
Have you tried the workaround outlined here? https://github.com/dotnet/runtime/issues/67782

**Comment by mfkl** [comment]
https://github.com/dotnet/runtime/issues/49772 might be relevant, you could try
```
<linker>
    <assembly fullname="System.Text.Json">
        <namespace fullname="System.Text.Json" preserve="all" />
        <namespace fullname="System.Text.Json.Serialization" preserve="all" />
        <namespace fullname="System.Text.Json.Serialization.Converters" preserve="all" />
    </assembly>
</linker>
```

**Reply by jakedowns** [reply]
thanks for the suggestion. looks like i'm still getting an exception tho after adding link.xml into my Assets folder with the contents above ^

![image](https://user-images.githubusercontent.com/1683122/211428705-8836a3de-4c9c-47f8-90c9-50b7b154642e.png)

i tried wrapping in a try/catch to see if i could get any more specific info on the error, but it didn't return much
![image](https://user-images.githubusercontent.com/1683122/211428816-9c60890f-c1f5-48b0-b4c0-ecdd1e36f72a.png)


my instinct is to try to breakpoint debug it, but i don't know if i can do that from android -> android studio -> unity -> visual studio ?

i tried disabling "strip engine code"

![image](https://user-images.githubusercontent.com/1683122/211429376-782b8a2f-a9fe-427e-9810-323df4eb6a25.png)

here are my other player settings
![image](https://user-images.githubusercontent.com/1683122/211429530-476d7ca1-9c70-4399-bb9f-8f0604bc4286.png)

**Reply by jakedowns** [reply]
these are the extension dlls in my Assets folder

![image](https://user-images.githubusercontent.com/1683122/211430843-fe141263-76df-4339-b825-474c0aa74a99.png)

**Reply by jakedowns** [reply]
there are some continued updates to this ongoing saga

here: https://github.com/Tyrrrz/YoutubeExplode/issues/639#issuecomment-1378092280
and here https://forum.unity.com/threads/il2cpp-build-time-improvements-seeking-feedback.1064135/page-3#post-8721021

**Comment by jakedowns** [comment]
![image](https://user-images.githubusercontent.com/1683122/212176266-eafff096-d2ca-4fc0-8477-c531cfa7d25f.png)

i used chatgpt to help me write a little python script that would synchronize the correct dll's from visual studios packages folder into the assets folder

https://gist.github.com/jakedowns/c2c2efc42c31d0e110f2cd885ba4d7b4

hitting this error now tho:
![image](https://user-images.githubusercontent.com/1683122/212176607-75cca003-1e3c-4caf-8316-0882ec4812e9.png)

**Reply by jakedowns** [reply]
elevating this to an issue https://github.com/Tyrrrz/YoutubeExplode/issues/680

**Comment by adenek** [comment, marked as answer]
Since the issue was closed I add my awnser here: 
YoutubeExplode do work well in unity for android but you need to do a bit of workaround.

1. Install nuget for unity (https://github.com/GlitchEnzo/NuGetForUnity)
Install youtube explode and the dependency via nuget
2. Remove dependency that already exist in unity (will have a report of that in the console)
3. Close unity and add this dependencie to your Packages/manifest.json
"com.system-community.systemruntimecompilerservicesunsafe": "https://github.com/system-community/SystemRuntimeCompilerServicesUnsafe.git?path=Assets/_Root#6.0.0"
4. Move all YoutubeExplode and dependencies dll to a Plugins folder
5. Add a link.xml document to avoid striping of the dll.
6. Make sure you using Unity 2022+ 7. Restart unity and compile
After that all work :)

**Reply by jakedowns** [reply]
thanks so much. i don't know what i had messed up in my multiple attempts, but starting from scratch and following these instructions seems to have solved it. 

it must've had something to do with me trying different package managers / manually moving dlls around

thanks for the tip.

btw, i got an error when i tried to add the CompilerServicesUnsafe to the manifest, but, i copied that one manually, and it seems to be ok

![image](https://user-images.githubusercontent.com/1683122/213037683-da9973e2-7e47-4854-b726-a71ceac1e654.png)
![image](https://user-images.githubusercontent.com/1683122/213037713-1ef95c1f-6871-41c3-a07c-4f0ea60722f2.png)


now i get to go learn how to solve this:
![image](https://user-images.githubusercontent.com/1683122/213037744-1bcd2342-c4ad-4122-8152-1c1a8cf99932.png)
